semantic-python/src/Language/Python/Core.hs

```diff
@@ -13,26 +15,35 @@ import           Data.Core as Core
 import           Data.Foldable
 import           Data.Name as Name
 import           GHC.Generics
+import           GHC.Records
+import qualified Data.Loc
 import qualified TreeSitter.Python.AST as Py
 import           TreeSitter.Span (Span)
 
+-- We don't want to commit to a particular representation of Core syntax,
+-- but there are commonalities that repeatedly crop up and that clog type
+-- signatures, so here's a constraint-kind alias for its capabilities.
+type CoreSyntax sig t = ( Member Core sig
+                        , Member Ann sig
+                        , Carrier sig t
+                        , Foldable t
+                        )
+
 class Compile py where
   -- FIXME: we should really try not to fail
-  compile :: (Member Core sig, Carrier sig t, Foldable t, MonadFail m) => py -> m (t Name)
+  compile :: (CoreSyntax sig t, MonadFail m) => py -> m (t Name)
 
   default compile :: (MonadFail m, Show py) => py -> m (t Name)
   compile = defaultCompile
 
-  compileCC :: (Member Core sig, Carrier sig t, Foldable t, MonadFail m) => py -> m (t Name) -> m (t Name)
+  compileCC :: (CoreSyntax sig t, MonadFail m) => py -> m (t Name) -> m (t Name)
 
-  default compileCC :: ( Member Core sig
-                       , Carrier sig t
-                       , Foldable t
-                       , MonadFail m
-                       )
-                    => py -> m (t Name) -> m (t Name)
+  default compileCC :: (CoreSyntax sig t, MonadFail m) => py -> m (t Name) -> m (t Name)
   compileCC py cc = (>>>) <$> compile py <*> cc
 
+locate :: (HasField "ann" syntax Span, CoreSyntax sig t) => syntax -> t a -> t a
+locate syn = Core.annAt (Data.Loc.fromTSSpan (getField @"ann" syn))
```

Comment:
@robrix had a number of ideas on how we could avoid having to `locate` every syntax datum manually. The only actually-good solution is to use open recursion; doing it without open recursion entails adding a new `compile’` method to `Compile` and the mental gyrations associated with keeping track of `compile` and `compile’`, which seemed to me more delicate than just calling `locate` manually.

I’m sure that _someday_ I’ll fall victim to the siren song of open recursion. When that day comes, I’ll kill `locate` and make it so location information is slurped automatically.

Reply:
Open recursion isn’t the only solution; e.g. we could give a global definition of `compile` (rather than a parameter) which does `locate` and calls into `compile'`. Then the rule is simple: never call `compile'` outside of `compile`. (Similar story for `compileCC`, of course.) This also has the benefit of being a pretty simple intermediary step towards open recursion.

I think we would only want to go with full open recursion if we actually had other per-input-term behaviours we’d want to layer in for only some executions and not others, e.g. tracing. I don’t think that’s the case here, which simplifies things a great deal.